File: week1_decomposition1/2_connected_components/connected_components.cpp
#include <iostream>
#include <vector>
#include <functional>

using std::vector;
using std::pair;

int number_of_components(vector<vector<int> > &adj) {
  int res = 0;
  vector<bool> visited(adj.size(),false);

  std::function<void(int)> dfs = [&](int i){
    for(int w: adj[i]){
      if (visited[w]==false){
        visited[w] = true;
        dfs(w);
      }
    }
  };
  for (int i=0;i<adj.size();++i){
    if(visited[i]==false){
      visited[i] = true;
      dfs(i);
      res += 1;
    }
  }
  return res;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  std::cout << number_of_components(adj);
}

File: week1_decomposition1/2_connected_components/connected_components.py
#Uses python3

import sys


def number_of_components(adj):
    result = 0
    visited = [False]*len(adj)

    def dfs(v):
        for w in adj[v]:
            if visited[w]==False:
                visited[w]=True
                dfs(w)

    for i in range(len(adj)):
        if visited[i]==False:
            visited[i] = True
            dfs(i)
            result += 1
    return result

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(number_of_components(adj))
